### PR TITLE
Optimize HIR map

### PR DESCRIPTION
**Pull request by Zoxc**
Builds on https://github.com/rust-lang/rust/pull/59042

cc @ljedrz
r? @eddyb

### CODE REVIEW

**Comment by Xanewok** [comment]
@rust-timer build 890cc87d1001f65c608c711d7a4d0d48129a43c1

**Comment by Zoxc** [comment]
Seems like this approach works well. There's still some slight memory regressions though.

File: src/librustc/hir/map/mod.rs [hunk 1]
```diff
@@ -160,6 +161,8 @@ impl Forest {
     }
 }
 
+pub(super) type HirMap<'hir> = [Vec<Option<IndexVec<ItemLocalId, Option<Entry<'hir>>>>>; 2];
```
Comment (Xanewok) [inline src/librustc/hir/map/mod.rs]: It takes a while to decipher what the structure looks like and why it looks this way - it'd be great if we could add a doc-comment here with an explanation

File: src/librustc/hir/map/mod.rs [hunk 2]
```diff
@@ -1034,6 +1039,24 @@ impl<'hir> Map<'hir> {
         attrs.unwrap_or(&[])
     }
 
+    /// Returns an iterator that yields all the hir ids in the map.
+    fn all_ids<'a>(&'a self) -> impl Iterator<Item = HirId> + 'a {
+        let map = &self.map;
+        let spaces = [DefIndexAddressSpace::Low, DefIndexAddressSpace::High].iter().cloned();
+        spaces.flat_map(move |space| {
```
Comment (Xanewok) [inline src/librustc/hir/map/mod.rs]: It'd be great to also add a comment here and/or rustfmt the function body

**Review by Xanewok — COMMENTED** [review]
IIUC this fixes a regression introduced in https://github.com/rust-lang/rust/pull/59042 and doesn't look like a *very* invasive change (limited only to one data structure) - should we not land it?

File: src/librustc/hir/map/mod.rs (hunk 1 above)
Comment (eddyb) [inline src/librustc/hir/map/mod.rs]: `hir::map::Map` *is* the HIR map. Can you rename this type alias to `HirEntryMap` or something?

**Review by eddyb — CHANGES_REQUESTED** [review]
This looks great, r=me modulo nits!

File: src/librustc/hir/map/mod.rs [hunk 3]
```diff
@@ -1154,35 +1194,21 @@ impl<'a, 'hir> NodesMatchingSuffix<'a, 'hir> {
...
-                Some(Node::ImplItem(n)) => n.name(),
-                Some(Node::Variant(n)) => n.name(),
-                Some(Node::Field(n)) => n.name(),
-                _ => continue,
-            };
-            if self.matches_names(self.map.get_parent(idx), name) {
-                return Some(idx)
-            }
-        }
+    fn matces_suffix(&self, hir: HirId) -> bool {
```
Comment (bjorn3) [inline src/librustc/hir/map/mod.rs]: *matches

File: src/librustc/hir/map/mod.rs [hunk 4]
```diff
@@ -1012,13 +1050,16 @@ impl<'hir> Map<'hir> {
...
+                                 -> impl Iterator<Item = NodeId> + 'a {
+        let nodes = NodesMatchingSuffix {
             map: self,
             item_name: parts.last().unwrap(),
             in_which: &parts[..parts.len() - 1],
-            idx: ast::CRATE_NODE_ID,
-        }
+        };
+
+        self.all_ids().filter(move |hir| nodes.matces_suffix(*hir)).map(move |hir| {
```
Comment (bjorn3) [inline src/librustc/hir/map/mod.rs]: *matches

**Comment by bjorn3** [comment]
I am a little confused about the order of comments on this PR: bors merges (1h ago) -> me reviewing it (few min ago) -> Zoxc gives r+ (3h ago).

**Comment by Zoxc** [comment]
@bjorn3 It's github new feature, time-travelling comments.

**Comment by bjorn3** [comment]
When looking at my notifications list, it said bors was the last commenter :)

**Comment by Zoxc** [comment]
You can see that I will post my comment in 3 hours. Github simply sorts comments by time!

**Comment by bjorn3** [comment]
On mobile it says 2 hours ago though.

**Comment by Zoxc** [comment]
Perf result: https://perf.rust-lang.org/compare.html?start=b92d360c6cf029bd98c154cb510ec9e11b39bad6&end=d7f4a864e13dc88d7a28c0b742da844af4e3abc4&stat=wall-time

**Pull request merged**